evennia/contrib/turnbattle/tb_magic.py

```diff
@@ -0,0 +1,1293 @@
+            caller.msg("The spell '%s' can only be cast on %i %s." % (spell_to_cast, spelldata["max_targets"], targplural))
+            return
+            
+        # Set up our candidates for targets
+        target_candidates = []
+        
+        # If spell targets 'any' or 'other', any object in caster's inventory or location
+        # can be targeted by the spell.
+        if spelldata["target"] in ["any", "other"]:
+            target_candidates = caller.location.contents + caller.contents
+        
+        # If spell targets 'anyobj', only non-character objects can be targeted.
+        if spelldata["target"] == "anyobj":
+            prefilter_candidates = caller.location.contents + caller.contents
+            for thing in prefilter_candidates:
+                if not thing.attributes.has("max_hp"): # Has no max HP, isn't a fighter
+                    target_candidates.append(thing)
+                    
+        # If spell targets 'anychar' or 'otherchar', only characters can be targeted.
+        if spelldata["target"] in ["anychar", "otherchar"]:
+            prefilter_candidates = caller.location.contents
+            for thing in prefilter_candidates:
+                if thing.attributes.has("max_hp"): # Has max HP, is a fighter
+                    target_candidates.append(thing)
+                    
+        # Now, match each entry in spell_targets to an object in the search candidates
+        matched_targets = []
+        for target in spell_targets:
+            match = caller.search(target, candidates=target_candidates)
+            matched_targets.append(match)
+        spell_targets = matched_targets
+                    
+        # If no target is given and the spell's target is 'self', set target to self
+        if len(spell_targets) == 0 and spelldata["target"] == "self":
+            spell_targets = [caller]
+                    
+        # Give error message if trying to cast an "other" target spell on yourself
+        if spelldata["target"] in ["other", "otherchar"]:
+            if caller in spell_targets:
+                caller.msg("You can't cast '%s' on yourself." % spell_to_cast)
+                return
+                
+        # Return if "None" in target list, indicating failed match
+        if None in spell_targets:
+            # No need to give an error message, as 'search' gives one by default.
+            return
+                
+        # Give error message if repeats in target list
+        if len(spell_targets) != len(set(spell_targets)):
+            caller.msg("You can't specify the same target more than once!")
+            return
+        
+        # Finally, we can cast the spell itself. Note that MP is not deducted here!
+        try:
+            spelldata["spellfunc"](caller, spell_to_cast, spell_targets, spelldata["cost"], **kwargs)
+        except Exception:
+            log_trace("Error in callback for spell: %s." % spell_to_cast)
+        """
```

Comment:
This should be moved to the end of the main docstring or made into a comment with `#`.